CORE-1316 Add compression to stream api

rabbitonweb opened the pull request:
## Overview
Based on previous @tymm work.
All messages that go through `TransportLayer.stream` are compressed by default, and decompressed when received on the other side

### Which JIRA issue does this PR relate to? If there is not a JIRA issue addressing this work, please create one now and add the link here.
CORE-1316

rabbitonweb commented:
@tymm s`stream ` is intended to be used *only* for bigger objects - like blocks

rabbitonweb commented:
What do you mean by not creating a thread pool?

tymm commented:
In `ByteStringOps.scala` I am creating a new fixed thread pool (with two threads) every time we compress or decompress (see line 60 and 109).
That is not really optimal due to the overhead of creating new threads.
The node should probably provide a suitable thread pool and pass it to `compressLZ4` and `decompressLZ4`. Maybe @sebtomba can help with that since he did the work on thread pool management in the node.

Can we really assume that blocks are bigger than 1MB most of the time?

odeac commented on shared/src/main/scala/coop/rchain/shared/ByteStringOps.scala:
```diff
@@ -0,0 +1,116 @@
+      override def run(): Unit = {
+        var lz4os: FramedLZ4CompressorOutputStream = null
+        try {
+          lz4os = new FramedLZ4CompressorOutputStream(pos, Parameters.DEFAULT)
```
why do you use a var and not initialize lz4os with `new FramedLZ4CompressorOutputStream(...)` in the first place?

rabbitonweb replied:
This is the most likely java'like copy-paste. There is an try-catch part, which could be exchanged for `Try`

tymm replied:
I wish I could have just copy-pasted this from somewhere ;)

I'm using a var because I want to close `lz4os` in the finally block in the case that `new FramedLZ4CompressorOutputStream(...)` explodes.

odeac replied:
if the c-tor explodes there is nothing to close

tymm replied:
Not necessarily.
For example `FramedLZ4CompressorOutputStream.close()` closes the output stream that gets passed to the constructor.
If the constructor explodes after the output stream is set to `this.out`, there is something to close.
After thinking about this it seems like line 44 and 92 are not needed.

odeac commented on shared/src/main/scala/coop/rchain/shared/ByteStringOps.scala:
```diff
@@ -0,0 +1,116 @@
+          while (count != -1) {
+            lz4os.write(data, 0, count)
+            count = is.read(data)
+          }
```
this can be extracted into a `transferAll(InputStream, OutputStream)` function and used twice

odeac commented on shared/src/main/scala/coop/rchain/shared/ByteStringOps.scala:
```diff
@@ -0,0 +1,116 @@
+        var lz4os: FramedLZ4CompressorOutputStream = null
+        try {
+          lz4os = new FramedLZ4CompressorOutputStream(pos, Parameters.DEFAULT)
+          val data  = new Array[Byte](8192)
```
why 8192?

tymm replied:
See for example this: https://stackoverflow.com/a/237495

By the way `java.io.BufferedReader` also uses 8192 as the default buffer size.

odeac commented on shared/src/main/scala/coop/rchain/shared/ByteStringOps.scala:
```diff
@@ -0,0 +1,116 @@
+    bytestring.get()
+  }
+
+  private def decompressLZ4(bs: ByteString): Option[ByteString] = {
```
this is mostly a copy+paste of the `compressLZ4` function. Can you avoid the duplication?

rabbitonweb replied:
not really though, there are differences worth not keeping them DRY

rabbitonweb commented:
@tymm I have not even noticed the executor. Why would we need to do it like that? Is is it due to one consumer and one producer working in parallel?

tymm commented:
@rabbitonweb as discussed on Discord we just need to make sure that there are two threads available per compression (or decompression). We should have a thread pool where `compress` and `decompress` can run on.
I just thought it was better to leave that part for you guys because I didn't want to mess with the different thread pools in the node.

tymm commented on shared/src/main/scala/coop/rchain/shared/Compression.scala:
```diff
@@ -0,0 +1,28 @@
+  }
+  def decompress(compressed: Array[Byte], decompressedLength: Int): Array[Byte] = {
+    val restored = new Array[Byte](decompressedLength)
+    decompressor.decompress(compressed, 0, restored, 0, decompressedLength);
```
`;` at the end of the line.

rabbitonweb replied:
java copy-paste :D

rabbitonweb commented:
thx @tymm for catching this! see my next commit

rabbitonweb commented:
@tymm I've fixed it and added test

```scala
    it("should compress effectivly when data is compressable (repeatable patterns)") {
      val word       = Array.fill(1024)((Random.nextInt(24) + 33).toByte)
      val ar         = Array.concat(List.fill(1024)(word): _*)
      val compressed = Compression.compress(ar)
      val ratio      = compressed.length.toDouble / ar.length.toDouble
      ar.length should be > compressed.length
      ratio shouldBe (0.005 +- 0.0001)
    }
```

rabbitonweb commented on shared/src/main/scala/coop/rchain/shared/Compression.scala:
```diff
@@ -0,0 +1,28 @@
+    val length              = compressor.compress(content, 0, content.length, compressed, 0, maxCompressedLength)
+    compressed.take(length)
+  }
+  def decompress(compressed: Array[Byte], decompressedLength: Int): Array[Byte] = {
```
This was not throwing exceptions, but I will add property based test just in case.

tymm replied:
```
Exception in thread "main" net.jpountz.lz4.LZ4Exception: Error decoding offset 1936282 of input buffer
	at net.jpountz.lz4.LZ4JNIFastDecompressor.decompress(LZ4JNIFastDecompressor.java:39)
```
That's the exception I saw.

rabbitonweb commented on shared/src/test/scala/coop/rchain/shared/CompressionSpec.scala:
```diff
@@ -0,0 +1,40 @@
+      }
+    }
+
+    it("should compress effectivly when data is compressable (repeatable patterns)") {
```
👍
